src/algebra/star/chsh.lean

```diff
@@ -0,0 +1,265 @@
+      simp only [mul_comm A₁ A₀, mul_comm B₁ B₀, mul_left_comm A₁ A₀, mul_left_comm B₁ B₀,
+        mul_left_comm B₀ A₀, mul_left_comm B₀ A₁, mul_left_comm B₁ A₀, mul_left_comm B₁ A₁],
+      norm_num,
+      simp only [mul_comm _ (2 : R), mul_comm _ (4 : R),
+        mul_left_comm _ (2 : R), mul_left_comm _ (4 : R)],
+      abel,
+      simp only [neg_mul_eq_neg_mul_symm, mul_one, int.cast_bit0, one_mul, int.cast_one,
+        gsmul_eq_mul, int.cast_neg],
+      simp only [←mul_assoc, ←add_assoc],
+      norm_num, },
+    have idem' : P = (1 / 4 : ℝ) • (P * P),
+    { have h : 4 * P = (4 : ℝ) • P := by simp [algebra.smul_def],
+      rw [idem, h, ←mul_smul],
+      norm_num, },
+    have sa : star P = P,
+    { dsimp [P],
+      simp only [star_add, star_sub, star_mul, star_bit0, star_one,
+        T.A₀_sa, T.A₁_sa, T.B₀_sa, T.B₁_sa, mul_comm B₀, mul_comm B₁], },
+    rw idem',
+    conv_rhs { congr, skip, congr, rw ←sa, },
+    convert smul_le_smul_of_nonneg (star_mul_self_nonneg : 0 ≤ star P * P) _,
+    { simp, },
+    { apply_instance, },
+    { norm_num, }, },
+  apply le_of_sub_nonneg,
+  simpa only [sub_add_eq_sub_sub, ←sub_add] using i₁,
+end
+
+/-!
+We now prove some rather specialized lemmas in preparation for the Tsirelson inequality,
+which we hide in a namespace as they are unlikely to be useful elsewhere.
+-/
+local notation `√2` := (real.sqrt 2 : ℝ)
+
+namespace tsirelson_inequality
+
+/-!
+We next need some lemmas about numerals in modules and algebras.
+The awkward appearance of both `•ℤ` and `•` seems unavoidable because later calculations by `abel`
+will introduce `•ℤ`.
+If anyone sees how to obtain these from general statements, please improve this!
+-/
+
+lemma two_gsmul_half_smul {α : Type*} [add_comm_group α] [module ℝ α] {X : α} :
+  2 •ℤ (2⁻¹ : ℝ) • X = X :=
+by { rw [gsmul_eq_smul_cast ℝ, ←mul_smul]; norm_num, }
+
+lemma neg_two_gsmul_half_smul {α : Type*} [add_comm_group α] [module ℝ α] {X : α} :
+  (-2) •ℤ (2⁻¹ : ℝ) • X = - X :=
+by { rw [gsmul_eq_smul_cast ℝ, ←mul_smul]; norm_num, }
+
+lemma smul_two {α : Type*} [ring α] [algebra ℝ α] {x : ℝ} :
+  x • (2 : α) = (2 * x) • 1 :=
+by { rw [mul_comm 2 x, mul_smul], simp, }
+
+lemma smul_four {α : Type*} [ring α] [algebra ℝ α] {x : ℝ} :
+  x • (4 : α) = (4 * x) • 1 :=
+by { rw [mul_comm 4 x, mul_smul], simp, }
+
+/-!
+Before proving Tsirelson's bound,
+we prepare some easy lemmas about √2.
+-/
+
+-- This calculation, which we need for Tsirelson's bound,
+-- defeated me. Thanks for the rescue from Shing Tak Lam!
+lemma tsirelson_inequality_aux : √2 * √2 ^ 3 = √2 * (2 * √2⁻¹ + 4 * (√2⁻¹ * 2⁻¹)) :=
+begin
+  ring,
+  rw [mul_assoc, inv_mul_cancel, real.sqrt_eq_rpow, ←real.rpow_nat_cast, ←real.rpow_mul],
+  { norm_num,
+    rw show (2 : ℝ) ^ (2 : ℝ) = (2 : ℝ) ^ (2 : ℕ), by { rw ←real.rpow_nat_cast, norm_num },
+    norm_num },
+  { norm_num, },
+  { norm_num, },
+end
+
+lemma sqrt_two_inv_sq : √2⁻¹ * √2⁻¹ = (2⁻¹ : ℝ) :=
+by { rw [←mul_inv'], norm_num, }
```

Comment:
Hmm, I was thinking that maybe a generalization of this could go in `data.real.sqrt`, but I guess `2` is special because it's a numeral?

Maybe `norm_num` needs some extensions to handle this (and some of the other lemmas above).

Reply:
Yes, and it's not even possible to write lemmas about `bit0` and `bit1` here, as we'd need to keep track of positivity of the numerals.

Reply:
@digama0 What's your opinion of lemmas like this?